<a href="https://colab.research.google.com/github/Tonry12/BSC_DPDM23/blob/main/Frequent_Patterns_(Association_Rules).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [3]:
import pandas as pd
import os

In [4]:
from google.colab import drive
drive.mount('/content/drive')
path = '/content/drive/MyDrive/DPDM2023'

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [5]:
data = pd.read_csv(os.path.join(path,'reduced_marketbasket.csv'),encoding='latin-1')

In [6]:
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  
...                  ...        ...         ...             ...  
541904  12/09/2011 12:50       0.85     12680.0          France  
541905  12/09/2011 12:50       2.10     12680.0          France  
541906  12/09/2011 12:50       4.15     12680.0          France  
541907  12/09/2011 12:50       4.15     12680.0          France  
541908  12/09/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

## แปลง data ให้อยู่ในรูปแบบ Transaction data ด้วย groupby.apply(list)

In [7]:
data.groupby('InvoiceNo')['StockCode'].apply(list)
#เราดึงมาแค่ รหัสinvoice โดยจะ groupby stockcode มาทำเป็น list

InvoiceNo
536365     [85123A, 71053, 84406B, 84029G, 84029E, 22752,...
536366                                        [22633, 22632]
536367     [84879, 22745, 22748, 22749, 22310, 84969, 226...
536368                          [22960, 22913, 22912, 22914]
536369                                               [21756]
                                 ...                        
C581484                                              [23843]
C581490                                       [22178, 23144]
C581499                                                  [M]
C581568                                              [21258]
C581569                                       [84978, 20979]
Name: StockCode, Length: 25900, dtype: object

In [8]:
transactions = data.groupby('InvoiceNo')['StockCode'].apply(list)

## Association Rules

In [9]:
!pip install apyori

  Preparing metadata (setup.py) ... done
  Created wheel for apyori: filename=apyori-1.1.2-py3-none-any.whl size=5954 sha256=56cb73f1353f35494babbefbba5f6d5044bf11191373c9bbd1fa51298344914f
  Stored in directory: /root/.cache/pip/wheels/c4/1a/79/20f55c470a50bb3702a8cb7c94d8ada15573538c7f4baebe2d
Successfully built apyori


In [10]:
from apyori import apriori

In [11]:
apriori(transactions,min_support = 0.065)
#มีอย่างน้อย 6.5% ที่ซื้อของซ้ำๆกันใน transation ก็ถือว่ามี pattern แล้ว

<generator object apriori at 0x7b63aa90b7d0>

In [12]:
list(apriori(transactions,min_support = 0.065))

[RelationRecord(items=frozenset({'22423'}), support=0.08386100386100385, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08386100386100385, lift=1.0)]),
 RelationRecord(items=frozenset({'47566'}), support=0.06586872586872587, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'47566'}), confidence=0.06586872586872587, lift=1.0)]),
 RelationRecord(items=frozenset({'85099B'}), support=0.08243243243243244, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B'}), confidence=0.08243243243243244, lift=1.0)]),
 RelationRecord(items=frozenset({'85123A'}), support=0.08671814671814672, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85123A'}), confidence=0.08671814671814672, lift=1.0)])]

In [13]:
frequentItemsets = list(apriori(transactions,min_support = 0.065))

In [14]:
len(frequentItemsets)

4

In [15]:
# อ่าน itemset1
frequentItemsets[0]
# มีแค่สินค้าชิ้นเดียว มี support 8.3% ที่ซื้อสินค้านี้
#ซึ่งไม่ได้อะไรมากเท่าไหร่

RelationRecord(items=frozenset({'22423'}), support=0.08386100386100385, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08386100386100385, lift=1.0)])

In [16]:
frequentItemsets[0][0]

frozenset({'22423'})

In [17]:
dir(frequentItemsets[0][0])

['__and__',
 '__class__',
 '__class_getitem__',
 '__contains__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__ne__',
 '__new__',
 '__or__',
 '__rand__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__ror__',
 '__rsub__',
 '__rxor__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__sub__',
 '__subclasshook__',
 '__xor__',
 'copy',
 'difference',
 'intersection',
 'isdisjoint',
 'issubset',
 'issuperset',
 'symmetric_difference',
 'union']

In [18]:
frequentItemsets[0][0].__len__()

# itemset นี้มีขนาดเท่ากับ 1

1

In [19]:
min_sup = 0.03

frequentItemsets = list(apriori(transactions,min_support = min_sup))
for i in frequentItemsets:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'85099B', '22386'}), support=0.03216216216216216, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B', '22386'}), confidence=0.03216216216216216, lift=1.0), OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335), OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])
yes
RelationRecord(items=frozenset({'22697', '22699'}), support=0.03027027027027027, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22697', '22699'}), confidence=0.03027027027027027, lift=1.0), OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947), OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])


มี 2 itemset ที่ผ่าน 0.03

RelationRecord(items=frozenset({'22386', '85099B'}),

support=0.03216216216216216,
<t>
<t>

คนที่ซื้อทั้ง 22386, 85099B จาก dataset ทั้งหมดมี 3.21%

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22386', '85099B'}), confidence=0.03216216216216216, lift=1.0),

คนที่ซื้อ 22386 แล้วค่อยซื้อ 85099B มีตั้ง 67.66%


OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335),


คนที่ซื้อ 85099B แล้วค่อยซื้อ 22386 มี 39.01%


OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])


In [20]:
set(data[(data['StockCode']=='22386') | (data['StockCode']=='85099B') | (data['StockCode']=='22699') | (data['StockCode']=='22697') ]['Description'])

{'GREEN REGENCY TEACUP AND SAUCER',
 'JUMBO BAG PINK POLKADOT',
 'JUMBO BAG RED RETROSPOT',
 'ROSES REGENCY TEACUP AND SAUCER '}

### quize 22386, 85099B คือสินค้าอะไร

In [21]:
set(data[data['StockCode']=='22386']['Description'])

{'JUMBO BAG PINK POLKADOT'}

In [22]:
set(data[data['StockCode']=='85099B']['Description'])

{'JUMBO BAG RED RETROSPOT'}

In [23]:
data[data['InvoiceNo']=='562950']

InvoiceNo StockCode                         Description  Quantity  \
297488    562950     22752        SET 7 BABUSHKA NESTING BOXES         2   
297489    562950     22755     SMALL PURPLE BABUSHKA NOTEBOOK         12   
297490    562950     22753     SMALL YELLOW BABUSHKA NOTEBOOK         12   
297491    562950     22754        SMALL RED BABUSHKA NOTEBOOK         12   
297492    562950     22757        LARGE RED BABUSHKA NOTEBOOK         12   
297493    562950     22758    LARGE PURPLE BABUSHKA NOTEBOOK          12   
297494    562950     22756     LARGE YELLOW BABUSHKA NOTEBOOK         12   
297495    562950     23310             BUBBLEGUM RING ASSORTED        36   
297496    562950     23280  FOLDING BUTTERFLY MIRROR HOT PINK         12   
297497    562950     22383              LUNCH BAG SUKI DESIGN         10   
297498    562950     22492             MINI PAINT SET VINTAGE         36   
297499    562950     71459      HANGING JAM JAR T-LIGHT HOLDER        12   
297500    562950     84945  MULTI COLOUR SILVER T-LIGHT HOLDER        12   
297501    562950     21175         GIN + TONIC DIET METAL SIGN        12   
297502    562950     23164                  REGENCY CAKE SLICE         4   
297503    562950     23245          SET OF 3 REGENCY CAKE TINS         4   
297504    562950     23171            REGENCY TEA PLATE GREEN         12   
297505    562950     22697     GREEN REGENCY TEACUP AND SAUCER         6   

             InvoiceDate  UnitPrice  CustomerID         Country  
297488  08/11/2011 09:56       8.50     15608.0  United Kingdom  
297489  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297490  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297491  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297492  08/11/2011 09:56       1.25     15608.0  United Kingdom  
297493  08/11/2011 09:56       1.25     15608.0  United Kingdom  
297494  08/11/2011 09:56       1.25     15608.0  United Kingdom  
297495  08/11/2011 09:56       0.42     15608.0  United Kingdom  
297496  08/11/2011 09:56       0.83     15608.0  United Kingdom  
297497  08/11/2011 09:56       1.65     15608.0  United Kingdom  
297498  08/11/2011 09:56       0.65     15608.0  United Kingdom  
297499  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297500  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297501  08/11/2011 09:56       2.55     15608.0  United Kingdom  
297502  08/11/2011 09:56       4.95     15608.0  United Kingdom  
297503  08/11/2011 09:56       4.95     15608.0  United Kingdom  
297504  08/11/2011 09:56       1.65     15608.0  United Kingdom  
297505  08/11/2011 09:56       2.95     15608.0  United Kingdom

# Homework 1 หา Frequent Item Set และ Association Rules
## ส่งก่อนจันทร์หน้าๆ
จัดกลุ่มตาม invoice
* ของช่วงเวลา Christmas & Newn Year (1 Dec - 31 Dec)
* ของประเทศในทวีป Europe
* ของช่วงเวลา ตรุษจีน ในทวีป Asia (15 Jan - 15 Feb)
* ของช่วงเวลา Christmas & Newn Year ในทวีป Europe (1 Dec - 31 Dec)
* ของวันวาเลนไทน์
* จัดกลุ่มตามประเทศและเดือน


In [49]:
Europe = data[(data['Country']!='Bahrain') | (data['Country']!='Hong Kong')| (data['Country']!='Singapore') | (data['Country']!='Japan') | (data['Country']!='Israel') | (data['Country']!='Lebanonl') | (data['Country']!='Saudi Arabia') | (data['Country']!='United Arab Emirates')]

In [50]:
Europe

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  
...                  ...        ...         ...             ...  
541904  12/09/2011 12:50       0.85     12680.0          France  
541905  12/09/2011 12:50       2.10     12680.0          France  
541906  12/09/2011 12:50       4.15     12680.0          France  
541907  12/09/2011 12:50       4.15     12680.0          France  
541908  12/09/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

In [55]:
tran_E = Europe.groupby('InvoiceNo')['Description'].apply(list).reset_index() #ทำการ group และทำให้ข้อมูลใน Description เป็นlist
tran_E2 = tran_E['Description'].apply(lambda x: [str(item) for item in x]) #ใช้ฟังก์ชัน lambda เพื่อแปลงสินค้าทุกชิ้นใน list ของ 'tr

In [58]:
tran_E2

0        [WHITE HANGING HEART T-LIGHT HOLDER, WHITE MET...
1        [HAND WARMER UNION JACK, HAND WARMER RED POLKA...
2        [ASSORTED COLOUR BIRD ORNAMENT, POPPY'S PLAYHO...
3        [JAM MAKING SET WITH JARS, RED COAT RACK PARIS...
4                               [BATH BUILDING BLOCK WORD]
                               ...                        
25895                        [PAPER CRAFT , LITTLE BIRDIE]
25896    [VICTORIAN GLASS HANGING T-LIGHT, ZINC T-LIGHT...
25897                                             [Manual]
25898                         [VICTORIAN SEWING BOX LARGE]
25899    [HANGING HEART JAR T-LIGHT HOLDER, 36 PENCILS ...
Name: Description, Length: 25900, dtype: object

In [63]:
frequentItemset_Europe = list(apriori(tran_E2, min_support=0.025))

for record in frequentItemset_Europe:
    if len(record.items) > 1:
        print('Yes')
        print(f"Items: {', '.join(record.items).lower()}")
        print(f"Support: {record.support:.4f} ({record.support*100:.2f}%)")

        for i, stat in enumerate(record.ordered_statistics, 1):
            items_base = 'None' if not stat.items_base else ', '.join(stat.items_base)
            items_add = ', '.join(stat.items_add)
            confidence = stat.confidence
            lift = stat.lift

            print(f"Rule {i}: ({items_base} -> {items_add} : Confidence={confidence:.4f}, Lift={lift:.4f})")
        print()


Yes
Items: green regency teacup and saucer, roses regency teacup and saucer 
Support: 0.0303 (3.03%)
Rule 1: (None -> GREEN REGENCY TEACUP AND SAUCER, ROSES REGENCY TEACUP AND SAUCER  : Confidence=0.0303, Lift=1.0000)
Rule 2: (GREEN REGENCY TEACUP AND SAUCER -> ROSES REGENCY TEACUP AND SAUCER  : Confidence=0.7417, Lift=17.1523)
Rule 3: (ROSES REGENCY TEACUP AND SAUCER  -> GREEN REGENCY TEACUP AND SAUCER : Confidence=0.7000, Lift=17.1523)

Yes
Items: jumbo bag red retrospot, jumbo bag pink polkadot
Support: 0.0322 (3.22%)
Rule 1: (None -> JUMBO BAG RED RETROSPOT, JUMBO BAG PINK POLKADOT : Confidence=0.0322, Lift=1.0000)
Rule 2: (JUMBO BAG PINK POLKADOT -> JUMBO BAG RED RETROSPOT : Confidence=0.6767, Lift=8.2090)
Rule 3: (JUMBO BAG RED RETROSPOT -> JUMBO BAG PINK POLKADOT : Confidence=0.3902, Lift=8.2090)

Yes
Items: jumbo bag red retrospot, jumbo shopper vintage red paisley
Support: 0.0264 (2.64%)
Rule 1: (None -> JUMBO BAG RED RETROSPOT, JUMBO SHOPPER VINTAGE RED PAISLEY : Confidence=0

In [82]:
import datetime

In [83]:
data['InvoiceDate']

0         12/01/2010 08:26
1         12/01/2010 08:26
2         12/01/2010 08:26
3         12/01/2010 08:26
4         12/01/2010 08:26
                ...       
541904    12/09/2011 12:50
541905    12/09/2011 12:50
541906    12/09/2011 12:50
541907    12/09/2011 12:50
541908    12/09/2011 12:50
Name: InvoiceDate, Length: 541909, dtype: object

In [97]:
data['InvoiceDate'] = pd.to_datetime(data['InvoiceDate'], format='%m/%d/%Y %H:%M')

In [98]:
data['Date']=data['InvoiceDate'].dt.date

In [100]:
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country        Date  
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom  2010-12-01  
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010-12-01  
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom  2010-12-01  
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010-12-01  
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom  2010-12-01  
...                    ...        ...         ...             ...         ...  
541904 2011-12-09 12:50:00       0.85     12680.0          France  2011-12-09  
541905 2011-12-09 12:50:00       2.10     12680.0          France  2011-12-09  
541906 2011-12-09 12:50:00       4.15     12680.0          France  2011-12-09  
541907 2011-12-09 12:50:00       4.15     12680.0          France  2011-12-09  
541908 2011-12-09 12:50:00       4.95     12680.0          France  2011-12-09  

[541909 rows x 9 columns]

In [101]:

data['Date']=pd.to_datetime(data.Date, format='%Y/%m/%d')

In [102]:
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country       Date  
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom 2010-12-01  
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom 2010-12-01  
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom 2010-12-01  
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom 2010-12-01  
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom 2010-12-01  
...                    ...        ...         ...             ...        ...  
541904 2011-12-09 12:50:00       0.85     12680.0          France 2011-12-09  
541905 2011-12-09 12:50:00       2.10     12680.0          France 2011-12-09  
541906 2011-12-09 12:50:00       4.15     12680.0          France 2011-12-09  
541907 2011-12-09 12:50:00       4.15     12680.0          France 2011-12-09  
541908 2011-12-09 12:50:00       4.95     12680.0          France 2011-12-09  

[541909 rows x 9 columns]

In [103]:
Asia = data[(data['Country']=='Bahrain') | (data['Country']=='Hong Kong')| (data['Country']=='Singapore') | (data['Country']=='Japan') | (data['Country']=='Israel') | (data['Country']=='Lebanonl') | (data['Country']=='Saudi Arabia') | (data['Country']=='United Arab Emirates')]

In [104]:
Asia

InvoiceNo StockCode                     Description  Quantity  \
9783      537218     85016   SET OF 6 VINTAGE NOTELETS KIT         6   
9784      537218     21506      FANCY FONT BIRTHDAY CARD,         12   
9785      537218     22716              CARD CIRCUS PARADE        12   
9786      537218     22983             CARD BILLBOARD FONT        12   
9787      537218     22030          SWALLOWS GREETING CARD        12   
...          ...       ...                             ...       ...   
530276   C580832     21731   RED TOADSTOOL LED NIGHT LIGHT        -2   
530277   C580832     23480  MINI LIGHTS WOODLAND MUSHROOMS       -15   
530278   C580832     22466  FAIRY TALE COTTAGE NIGHT LIGHT       -17   
530279   C580832     23084              RABBIT NIGHT LIGHT        -7   
530280   C580832     22621      TRADITIONAL KNITTING NANCY        -1   

               InvoiceDate  UnitPrice  CustomerID Country       Date  
9783   2010-12-05 15:46:00       2.55     12763.0   Japan 2010-12-05  
9784   2010-12-05 15:46:00       0.42     12763.0   Japan 2010-12-05  
9785   2010-12-05 15:46:00       0.42     12763.0   Japan 2010-12-05  
9786   2010-12-05 15:46:00       0.42     12763.0   Japan 2010-12-05  
9787   2010-12-05 15:46:00       0.42     12763.0   Japan 2010-12-05  
...                    ...        ...         ...     ...        ...  
530276 2011-12-06 11:40:00       1.45     12753.0   Japan 2011-12-06  
530277 2011-12-06 11:40:00       3.29     12753.0   Japan 2011-12-06  
530278 2011-12-06 11:40:00       1.65     12753.0   Japan 2011-12-06  
530279 2011-12-06 11:40:00       1.79     12753.0   Japan 2011-12-06  
530280 2011-12-06 11:40:00       1.45     12753.0   Japan 2011-12-06  

[1269 rows x 9 columns]

In [111]:
#ของช่วงเวลา ตรุษจีน ในทวีป Asia (15 Jan - 15 Feb)
Jan_1 = data[(data['Date'].dt.month == 1) & (data['Date'].dt.day >= 15) & (data['Date'].dt.day <= 31)]
#ปริ้นเพื่อตรวจสอบ
print(Jan_1)

      InvoiceNo StockCode                          Description  Quantity  \
58345    541238     21115                ROSE CARAVAN DOORSTOP         1   
58346    541238     22957  SET 3 PAPER VINTAGE CHICK PAPER EGG         1   
58347    541238     22758     LARGE PURPLE BABUSHKA NOTEBOOK           1   
58348    541238     20750             RED RETROSPOT MINI CASES         2   
58349    541238     22722    SET OF 6 SPICE TINS PANTRY DESIGN         1   
...         ...       ...                                  ...       ...   
77623    542745     22138        BAKING SET 9 PIECE RETROSPOT         13   
77624    542746     21034         REX CASH+CARRY JUMBO SHOPPER         2   
77625   C542748     22969         HOMEMADE JAM SCENTED CANDLES       -12   
77626   C542753     22649         STRAWBERRY FAIRY CAKE TEAPOT        -1   
77627    542759     22649         STRAWBERRY FAIRY CAKE TEAPOT         1   

              InvoiceDate  UnitPrice  CustomerID         Country       Date  
58345 201

In [113]:
Feb_1 = data[(data['Date'].dt.month == 2) & (data['Date'].dt.day >= 1) & (data['Date'].dt.day <= 15)]
print(Feb_1)

      InvoiceNo StockCode                        Description  Quantity  \
77628    542776     21671       RED SPOT CERAMIC DRAWER KNOB        48   
77629    542776     21668     RED STRIPE CERAMIC DRAWER KNOB        12   
77630    542776     21670      BLUE SPOT CERAMIC DRAWER KNOB        12   
77631    542776    17013D       ORIGAMI ROSE INCENSE IN TUBE        96   
77632    542776     17096  ASSORTED LAQUERED INCENSE HOLDERS        48   
...         ...       ...                                ...       ...   
91285    544089    15056N          EDWARDIAN PARASOL NATURAL         3   
91286    544089    84406B     CREAM CUPID HEARTS COAT HANGER         3   
91287    544089     22087           PAPER BUNTING WHITE LACE         3   
91288    544089     22631           CIRCUS PARADE LUNCH BOX          2   
91289    544089     22630               DOLLY GIRL LUNCH BOX         2   

              InvoiceDate  UnitPrice  CustomerID         Country       Date  
77628 2011-02-01 08:23:00       1

In [114]:
Asia_date = pd.concat([Jan_1,Feb_1])

In [105]:
trans_Asia = Asia.groupby('InvoiceNo')['Description'].apply(list).reset_index() #ทำการ group และทำให้ข้อมูลใน Description เป็นlist
trans_Asia3 = trans_Asia['Description'].apply(lambda x: [str(item) for item in x]) #ใช้ฟังก์ชัน lambda เพื่อแปลงสินค้าทุกชิ้นใน list ของ 'transactions' เป็นข้อความ (string)


In [68]:
frequentItemset_Asia = list(apriori(trans_Asia3, min_support=0.3))

for record in frequentItemset_Asia:
    if len(record.items) > 1:
        print('Yes')
        print(f"Items: {', '.join(record.items).lower()}")
        print(f"Support: {record.support:.4f} ({record.support*100:.2f}%)")

        for i, stat in enumerate(record.ordered_statistics, 1):
            items_base = 'None' if not stat.items_base else ', '.join(stat.items_base)
            items_add = ', '.join(stat.items_add)
            confidence = stat.confidence
            lift = stat.lift

            print(f"Rule {i}: ({items_base} -> {items_add} : Confidence={confidence:.4f}, Lift={lift:.4f})")
        print()
